File: Natural-Language-Processing/NLP-DecisionTree.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 21 09:44:49 2017

@author: adhingra
"""
# Natural Lanugage Processing
# using Decision Tree

# importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# importing dataset
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

# Cleaning the texts
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus = []
for i in range(0,1000):
    review = re.sub('[^a-zA-Z]', ' ' , dataset['Review'][i])
    review = review.lower()
    review = review.split()
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

# Fitting Bag of words model to the corpus
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1400)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:,1].values

# Splitting dataset into training and test sets
from sklearn.cross_validation import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Fitting Decision Tree Classification Model to the dataset
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train,y_train)

# predicting test set results
y_pred = classifier.predict(X_test)

# making confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)

# getting accuracy, preicison, recall , f1 score and classification report
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report
precision = precision_score(y_test,y_pred)
recall = recall_score(y_test,y_pred)
f1 = f1_score(y_test,y_pred)
accuracy = accuracy_score(y_test,y_pred)
classificationreport = classification_report(y_test,y_pred)

File: Natural-Language-Processing/NLP-DecisionTree.R
# Natural Language Processing 
# using Decision Tree Classification Model

# importing dataset
dataset = read.delim('Restaurant_Reviews.tsv', sep = '\t', quote = '', stringsAsFactors = FALSE)

# cleaning the texts
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset$Review))
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, content_transformer(tolower))
corpus = tm_map(corpus, removeWords, stopwords())
corpus = tm_map(corpus, stemDocument)
corpus = tm_map(corpus, stripWhitespace)

# fitting bag of words model
dtm = DocumentTermMatrix(corpus)
dataset_dtm = removeSparseTerms(dtm, 0.999)
dataset_dtm = as.data.frame(as.matrix(dataset_dtm))
dataset_dtm$Liked = dataset$Liked

# encoding the dependent variable as factor
dataset_dtm$Liked = factor(x = dataset_dtm$Liked, levels = c(0,1))

# splitting the dataset into training and test sets
library(caTools)
split = sample.split(dataset_dtm$Liked , SplitRatio = 0.8)
training_set = subset(dataset_dtm, split == TRUE)
test_set = subset(dataset_dtm, split == FALSE)

# fitting decision tree classification model to the new dataset
library(rpart)
classifier = rpart(formula = Liked ~ ., data = training_set)

# predicting test set results
y_pred = predict(classifier, newdata = test_set[,1:691], type = 'class')

# making confusion matrix
cm = table(test_set[,692],y_pred)

# getting accuracy, precision, f1 and recall score
library(caret)
accuracy = (cm[4] + cm[1]) / (cm[1] + cm[2] + cm[3] + cm[4])
precision_score = cm[4] / (cm[4] + cm[3])
recall_score = cm[4] / (cm[4] + cm[2])
f1 = (2*precision_score*recall_score)/(precision_score+recall_score)
